common/font.go

```diff
@@ -145,3 +146,154 @@ func (f *Font) Render(text string) Texture {
 	return NewTextureSingle(imObj)
 }
+
+// generateFontAtlas generates the font atlas for this given font, using the first `c` Unicode characters.
+func (f *Font) generateFontAtlas(c int) FontAtlas {
+	atlas := FontAtlas{
+		XLocation: make([]float32, c),
+		YLocation: make([]float32, c),
+		Width:     make([]float32, c),
+		Height:    make([]float32, c),
+	}
+
+	var (
+		int26Width  fixed.Int26_6
+		int26Height fixed.Int26_6
+
+		totalHeight fixed.Int26_6
+
+		fupe        = fixed.Int26_6(f.TTF.FUnitsPerEm())
+		maxYBearing = fixed.Int26_6(0)
+
+		currentX float32
+		maxX     float32
+
+		totalString string // TODO; string is immutable, so this is relatively inefficient; see also where we use `totalString += string(char)`
+		subString   string
+
+		drawCurY int
+	)
+
+	// The "full image"
+	nrgba := image.NewNRGBA(image.Rect(0, 0, 1200, 5000)) // way too big; hopefully
+
+	for i := 0; i < c; i++ {
+		char := rune(i)
+		totalString += string(char)
+		subString += string(char)
+		atlas.XLocation[char] = currentX
+		atlas.YLocation[char] = float32(drawCurY)
+
+		idx := f.TTF.Index(char)
+		hm := f.TTF.HMetric(fupe, idx)
+		vm := f.TTF.VMetric(fupe, idx)
+
+		g := truetype.GlyphBuf{}
+		err := g.Load(f.TTF, fupe, idx, font.HintingNone)
+		if err != nil {
+			log.Println("Error creating font atlas:", err)
+			return atlas
+		}
+
+		int26Width += hm.AdvanceWidth
+
+		//atlas.Width[char] = float32(hm.AdvanceWidth * fixed.Int26_6(f.Size) / fupe)
+		atlas.Width[char] = float32(g.AdvanceWidth * fixed.Int26_6(f.Size) / fupe)
+
+		currentX = float32(int26Width * fixed.Int26_6(f.Size) / fupe)
+		if currentX > maxX {
+			maxX = currentX
+		}
+
+		//yB := vm.TopSideBearing
+		yB := vm.TopSideBearing - g.Bounds.Min.Y
+		//yB := g.Bounds.Max.Y - g.Bounds.Min.Y
+		if yB > maxYBearing {
+			maxYBearing = yB
+		}
+		atlas.Height[char] = float32(yB * fixed.Int26_6(f.Size) / fupe)
+
+		if int(int26Width*fixed.Int26_6(f.Size)/fupe) > 1024 {
+			// Now let's draw these chars!
+			subimg := f.RenderNRGBA(subString)
+			// TODO: optimize this, because `f.RenderNRGBA` also allocates a new image, while we could be drawing to this one directly.
+			draw.Draw(nrgba, image.Rect(0, drawCurY, subimg.Bounds().Max.X, drawCurY+subimg.Bounds().Max.Y), subimg, image.ZP, draw.Src)
+			drawCurY += subimg.Bounds().Max.Y
+
+			int26Height += maxYBearing
+			totalHeight += maxYBearing
+			maxYBearing = fixed.Int26_6(0)
+
+			subString = ""
+			int26Width = 0
+			currentX = 0
+		}
+	}
+
+	// Draw the last line as well
+	subimg := f.RenderNRGBA(subString)
+	// TODO: optimize this, because `f.RenderNRGBA` also allocates a new image, while we could be drawing to this one directly.
+	draw.Draw(nrgba, image.Rect(0, drawCurY, subimg.Bounds().Max.X, drawCurY+subimg.Bounds().Max.Y), subimg, image.ZP, draw.Src)
+	drawCurY += subimg.Bounds().Max.Y
+
+	int26Height += maxYBearing
+	totalHeight += maxYBearing
+	maxYBearing = fixed.Int26_6(0)
+
+	subString = ""
+	int26Width = 0
+	currentX = 0
```

Comment:
These three lines also make no sense here (even though they make sense within the loop).